cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -639,8 +686,14 @@ private void addSystemArtifact(SystemArtifactInfo systemArtifactInfo) throws IOE
    *
    * @param artifactId the artifact to delete
    * @throws IOException if there was some IO error deleting the artifact
+   * @throws UnauthorizedException if the current user is not authorized to delete the artifact. To delete an artifact,
+   *                               a user needs {@link Action#ADMIN} permission on the artifact.
    */
-  public void deleteArtifact(Id.Artifact artifactId) throws IOException {
+  public void deleteArtifact(Id.Artifact artifactId) throws Exception {
+    authorizerInstantiatorService.get().enforce(artifactId.toEntityId(), SecurityRequestContext.toPrincipal(),
+                                                Action.ADMIN);
+    // revoke all privileges on the artifact
+    authorizerInstantiatorService.get().revoke(artifactId.toEntityId());
```

Comment:
As a rule, we're taking the safe route until we can perform delete as a single transaction. Don't want to land up in a situation where delete succeeds but revoke fails, since that would mean if someone creates the deleted artifact again, there are some default permissions on it.

So for now, if delete fails, it is likely that the user deleting it lost permissions to delete it in future. But in that scenario, we expect a CDAP 'admin' or another user who has admin privileges on the entity to clean things up.